File: python/medium/Sum_Root_to_Leaf_Numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = 0
        
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def next_node(node, num):
            if node.left is None and node.right is None:
                self.ans+=num*10+node.val
                return;
            num*=10
            if node.left is not None:
                next_node(node.left, num+node.val)
            if node.right is not None:
                next_node(node.right, num+node.val)
        next_node(root, 0)
        return self.ans

File: cpp/medium/Sum_Root_to_Leaf_Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int sumNumbers(TreeNode* root) {
        if (root == nullptr) return 0;
        next_node(root, 0);
        return ans;
    }
    void next_node(TreeNode* node, int num) {
        num = num * 10 + node->val;
        if (node->left == nullptr && node->right == nullptr) ans+=num;
        if (node->left !=nullptr) next_node(node->left, num);
        if (node->right !=nullptr) next_node(node->right, num);
    }
};
